File: tests/db/support/clickhouse.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { splitSqlStatements } from '@shared/domain/sql-split';
import type { ResolvedConnectionConfig } from '@shared/protocol/engine-ops';
import { ClickHouseContainer, type StartedClickHouseContainer } from '@testcontainers/clickhouse';
import { resolveDockerHost } from './docker';

resolveDockerHost();

// P36 D34: 26.3 is the current LTS with the longest remaining support window.
const IMAGE = 'clickhouse/clickhouse-server:26.3';
// The container's own admin user (CLICKHOUSE_USER/CLICKHOUSE_PASSWORD) — the official image
// grants it full access, which is what lets it CREATE USER below. Never returned as `config`.
const ADMIN_USER = 'kira_admin';
const ADMIN_PASSWORD = 'kira';
const DATABASE = 'kira_test';
const USERNAME = 'kira';
const PASSWORD = 'kira';
export const READONLY_USERNAME = 'kira_ro';
const READONLY_PASSWORD = 'kira';
const STARTUP_TIMEOUT_MS = 120_000;
const SEED_SQL_PATH = resolve(__dirname, '../fixtures/0010_clickhouse_seed.sql');

export interface ClickHouseFixture {
  container: StartedClickHouseContainer;
  /** The unprivileged `kira` user — SELECT/INSERT/ALTER DELETE on kira_test.* (D35). */
  config: ResolvedConnectionConfig;
  /** The `kira_ro` user — SELECT only; scenario 43's server-side readonly-enforcement target. */
  readOnlyConfig: ResolvedConnectionConfig;
  uri: string;
  /** The container's own HTTP base URL — for a spec file's own side `@clickhouse/client` (system
   *  table assertions, admin-only DDL a fixed grant set can't do). */
  baseUrl: string;
  adminUsername: string;
  adminPassword: string;
  database: string;
  stop(): Promise<void>;
}

// One container per test process, same discipline as every other support/*.ts fixture (§11b).
let memoized: Promise<ClickHouseFixture> | null = null;

export function startClickHouse(): Promise<ClickHouseFixture> {
  if (!memoized) memoized = start();
  return memoized;
}

// The HTTP interface has no multi-statement exec (unlike mariadb's importFile or node-postgres's
// multi-statement query) — each statement in the seed file is sent as its own request, split the
// same quote/comment-aware way the app's own console splits a "Run all" batch.
async function runStatements(
  baseUrl: string,
  user: string,
  password: string,
  database: string,
  sql: string,
): Promise<void> {
  const clickhouseModule = await import('@clickhouse/client');
  const client = clickhouseModule.createClient({
    url: baseUrl,
    username: user,
    password,
    database,
  });
  try {
    for (const stmt of splitSqlStatements(sql)) {
      await client.command({ query: stmt.text });
    }
  } finally {
    await client.close();
  }
}

async function start(): Promise<ClickHouseFixture> {
  // F47: the preset's own Wait.forHttp('/') predicate on "Ok.\n" and its nofile ulimit bump are
  // exactly what a hand-rolled GenericContainer would get wrong here — mirrors support/mysql.ts's
  // and support/mariadb.ts's own move onto the equivalent presets.
  const container = await new ClickHouseContainer(IMAGE)
    .withDatabase(DATABASE)
    .withUsername(ADMIN_USER)
    .withPassword(ADMIN_PASSWORD)
    .withStartupTimeout(STARTUP_TIMEOUT_MS)
    .start();

  const host = container.getHost();
  const port = container.getHttpPort(); // D10/F2: the client is HTTP-only, never the 9000 native port.
  const baseUrl = `http://${host}:${port}`;
  const uri = `clickhouse://${USERNAME}:${PASSWORD}@${host}:${port}/${DATABASE}`;

  const seedSql = readFileSync(SEED_SQL_PATH, 'utf8');
  await runStatements(baseUrl, ADMIN_USER, ADMIN_PASSWORD, DATABASE, seedSql);

  // D35: two unprivileged users, created only after the seed lands — the same root-seeds/
  // app-user-connects split support/mysql.ts's own comment explains (§6e), which is what makes
  // scenario 7's cancel assertion and scenario 43's readonly assertion meaningful: neither is a
  // superuser connection.
  await runStatements(
    baseUrl,
    ADMIN_USER,
    ADMIN_PASSWORD,
    DATABASE,
    `
    CREATE USER IF NOT EXISTS ${USERNAME} IDENTIFIED WITH plaintext_password BY '${PASSWORD}';
    GRANT SELECT, INSERT, ALTER DELETE ON ${DATABASE}.* TO ${USERNAME};
    GRANT SELECT ON system.* TO ${USERNAME};
    CREATE USER IF NOT EXISTS ${READONLY_USERNAME} IDENTIFIED WITH plaintext_password BY '${READONLY_PASSWORD}';
    GRANT SELECT ON ${DATABASE}.* TO ${READONLY_USERNAME};
    GRANT SELECT ON system.* TO ${READONLY_USERNAME};
    `,
  );

  const now = new Date().toISOString();
  const config: ResolvedConnectionConfig = {
    id: 'test-clickhouse',
    sortOrder: 0,
    createdAt: now,
    updatedAt: now,
    name: 'Test ClickHouse',
    kind: 'clickhouse',
    color: 'orange',
    mode: 'fields',
    readOnly: false,
    host,
    port,
    database: DATABASE,
    username: USERNAME,
    uri: null,
    options: {},
    password: PASSWORD,
  };
  const readOnlyConfig: ResolvedConnectionConfig = {
    ...config,
    id: 'test-clickhouse-ro',
    name: 'Test ClickHouse (read-only)',
    readOnly: true,
    username: READONLY_USERNAME,
    password: READONLY_PASSWORD,
  };

  return {
    container,
    config,
    readOnlyConfig,
    uri,
    baseUrl,
    adminUsername: ADMIN_USER,
    adminPassword: ADMIN_PASSWORD,
    database: DATABASE,
    async stop() {
      // Playwright's workers:1 config runs every UI spec file sequentially in the same worker
      // process, sharing this module's state (mirrors support/mysql.ts:196-201's own reset).
      memoized = null;
      await container.stop();
    },
  };
}

File: apps/kira-studio/internal/adapters/testsupport/clickhouse.go
package testsupport

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/testcontainers/testcontainers-go"
	tcclickhouse "github.com/testcontainers/testcontainers-go/modules/clickhouse"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// ClickHouseFixture is support/clickhouse.ts's ClickHouseFixture.
type ClickHouseFixture struct {
	Config         model.ResolvedConnectionConfig
	ReadOnlyConfig model.ResolvedConnectionConfig
	BaseURL        string
	container      testcontainers.Container
}

var clickhouseMemo fixture[ClickHouseFixture]

const (
	clickhouseImage      = "clickhouse/clickhouse-server:26.3"
	clickhouseAdminUser  = "kira_admin"
	clickhouseAdminPass  = "kira"
	clickhouseDatabase   = "kira_test"
	clickhouseUsername   = "kira"
	clickhousePassword   = "kira"
	clickhouseROUsername = "kira_ro"
	clickhouseROPassword = "kira"
)

// StartClickHouse is support/clickhouse.ts's startClickHouse. See fixture.go's own doc comment for
// why termination is never wired to t.Cleanup (B15).
func StartClickHouse(t *testing.T) *ClickHouseFixture {
	t.Helper()
	if !IsDockerAvailable() {
		t.Skip(DockerUnavailableMessage)
	}
	fixture, err := clickhouseMemo.get(startClickHouse)
	if err != nil {
		t.Fatalf("clickhouse container: %v", err)
	}
	return fixture
}

// StopClickHouse terminates the memoized container, if one was ever started. Call once, from the
// test binary's own TestMain, after m.Run() returns — never from an individual test.
func StopClickHouse() {
	clickhouseMemo.stop(func(f *ClickHouseFixture) { _ = f.container.Terminate(context.Background()) })
}

func startClickHouse() (*ClickHouseFixture, error) {
	ctx := context.Background()

	// Without CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT=1, the official image's own bootstrap grants
	// kira_admin plain GRANT ALL ON *.* — which excludes ACCESS MANAGEMENT (CREATE USER, GRANT,
	// ...) — so the CREATE USER statements below would fail with ACCESS_DENIED.
	container, err := tcclickhouse.Run(ctx, ImageFor("clickhouse", clickhouseImage),
		tcclickhouse.WithDatabase(clickhouseDatabase),
		tcclickhouse.WithUsername(clickhouseAdminUser),
		tcclickhouse.WithPassword(clickhouseAdminPass),
		testcontainers.WithEnv(map[string]string{"CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT": "1"}),
	)
	if err != nil {
		return nil, err
	}

	host, err := container.Host(ctx)
	if err != nil {
		return nil, err
	}
	port, err := container.MappedPort(ctx, "8123/tcp")
	if err != nil {
		return nil, err
	}
	baseURL := fmt.Sprintf("http://%s:%s", host, port.Port())

	seedPath := filepath.Join(repoRoot(), "packages", "db-fixtures", "fixtures", "0010_clickhouse_seed.sql")
	seedSQL, err := os.ReadFile(seedPath)
	if err != nil {
		return nil, err
	}
	if err := runClickHouseStatements(ctx, baseURL, clickhouseAdminUser, clickhouseAdminPass, clickhouseDatabase, string(seedSQL)); err != nil {
		return nil, err
	}

	// D35: two unprivileged users, created only after the seed lands — the same root-seeds/
	// app-user-connects split support/mysql.go's own comment explains, which is what makes the
	// cancel assertion and the read-only assertion meaningful: neither is a superuser connection.
	// P26 §3.1(3): kira's own ALTER carries CREATE TABLE and DROP TABLE too (ClickHouse's ALTER
	// privilege does not itself cover object creation/deletion) — needed for the DDL round-trip
	// scratch table this phase's own tests create and drop over this same principal.
	grants := fmt.Sprintf(`
		CREATE USER IF NOT EXISTS %[1]s IDENTIFIED WITH plaintext_password BY '%[2]s';
		GRANT SELECT, INSERT, ALTER, CREATE TABLE, DROP TABLE ON %[3]s.* TO %[1]s;
		GRANT SELECT ON system.* TO %[1]s;
		GRANT SELECT ON default.* TO %[1]s;
		CREATE USER IF NOT EXISTS %[4]s IDENTIFIED WITH plaintext_password BY '%[5]s';
		GRANT SELECT ON %[3]s.* TO %[4]s;
		GRANT SELECT ON system.* TO %[4]s;
		GRANT SELECT ON default.* TO %[4]s;
	`, clickhouseUsername, clickhousePassword, clickhouseDatabase, clickhouseROUsername, clickhouseROPassword)
	if err := runClickHouseStatements(ctx, baseURL, clickhouseAdminUser, clickhouseAdminPass, clickhouseDatabase, grants); err != nil {
		return nil, err
	}

	now := time.Now().UTC().Format(time.RFC3339Nano)
	portNum := int(port.Num())
	config := model.ResolvedConnectionConfig{
		ID: "test-clickhouse", SortOrder: 0, CreatedAt: now, UpdatedAt: now,
		Name: "Test ClickHouse", Kind: "clickhouse", Color: "orange", Mode: "fields", ReadOnly: false,
		Host: &host, Port: &portNum, Database: Strp(clickhouseDatabase), Username: Strp(clickhouseUsername),
		Options: map[string]any{}, Password: Strp(clickhousePassword),
	}
	readOnlyConfig := config
	readOnlyConfig.ID = "test-clickhouse-ro"
	readOnlyConfig.Name = "Test ClickHouse (read-only)"
	readOnlyConfig.ReadOnly = true
	readOnlyConfig.Username = Strp(clickhouseROUsername)
	readOnlyConfig.Password = Strp(clickhouseROPassword)

	return &ClickHouseFixture{Config: config, ReadOnlyConfig: readOnlyConfig, BaseURL: baseURL, container: container}, nil
}

// AdminStatements runs sql as ClickHouse's own admin user (kira_admin) — for a matrix case's own
// runtime role/grant statements, reusing the same statement splitter and HTTP path the fixture's
// own seed step already goes through.
func AdminStatements(t *testing.T, f *ClickHouseFixture, sql string) {
	t.Helper()
	if err := runClickHouseStatements(context.Background(), f.BaseURL, clickhouseAdminUser, clickhouseAdminPass, clickhouseDatabase, sql); err != nil {
		t.Fatalf("AdminStatements: %v", err)
	}
}

// firstTopLevelSemicolon mirrors sqlite/query.go's own scanner (comments and quoted literals are
// skipped over so a semicolon inside a string value never counts) — generalized here to find every
// boundary rather than just the first, since this fixture's own job (unlike the adapter's B9 guard)
// is to split a whole seed file into individually-POSTed statements: the ClickHouse HTTP interface
// has no multi-statement exec at all, unlike mariadb's importFile or node-postgres's own
// multi-statement query.
func splitClickHouseStatements(sql string) []string {
	const (
		normal = iota
		lineComment
		blockComment
		single
		backtick
	)
	state := normal
	var statements []string
	start := 0
	for i := 0; i < len(sql); i++ {
		c := sql[i]
		switch state {
		case normal:
			switch {
			case c == '-' && i+1 < len(sql) && sql[i+1] == '-':
				state = lineComment
				i++
			case c == '/' && i+1 < len(sql) && sql[i+1] == '*':
				state = blockComment
				i++
			case c == '\'':
				state = single
			case c == '`':
				state = backtick
			case c == ';':
				statements = append(statements, strings.TrimSpace(sql[start:i]))
				start = i + 1
			}
		case lineComment:
			if c == '\n' {
				state = normal
			}
		case blockComment:
			if c == '*' && i+1 < len(sql) && sql[i+1] == '/' {
				state = normal
				i++
			}
		case single:
			if c == '\\' {
				i++
			} else if c == '\'' {
				state = normal
			}
		case backtick:
			if c == '`' {
				state = normal
			}
		}
	}
	if strings.TrimSpace(sql[start:]) != "" {
		statements = append(statements, strings.TrimSpace(sql[start:]))
	}
	out := statements[:0]
	for _, s := range statements {
		if s != "" {
			out = append(out, s)
		}
	}
	return out
}

func runClickHouseStatements(ctx context.Context, baseURL, user, password, database, sql string) error {
	client := &http.Client{Timeout: 30 * time.Second}
	for _, stmt := range splitClickHouseStatements(sql) {
		u := baseURL + "/?database=" + database
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, strings.NewReader(stmt))
		if err != nil {
			return err
		}
		req.Header.Set("X-ClickHouse-User", user)
		req.Header.Set("X-ClickHouse-Key", password)
		resp, err := client.Do(req)
		if err != nil {
			return err
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			return fmt.Errorf("clickhouse seed statement failed (status %d): %s\nstatement: %s", resp.StatusCode, body, stmt)
		}
	}
	return nil
}
